codespan-reporting/src/term/renderer.rs

```diff
@@ -469,12 +470,17 @@ impl<'writer, 'config> Renderer<'writer, 'config> {
         range: Range<usize>,
         message: &str,
     ) -> io::Result<()> {
-        let space_len = self.config.width(&source[..range.start]);
+        let space_len = self
+            .config
+            .width(slice_at_char_boundaries(source, 0..range.start));
         write!(self, " {space: >width$}", space = "", width = space_len)?;
         self.set_color(self.styles().label(severity, label_style))?;
         // We use `usize::max` here to ensure that we print at least one
         // label character - even when we have a zero-length span.
-        for _ in 0..usize::max(self.config.width(&source[range.clone()]), 1) {
+        for _ in 0..usize::max(
+            self.config.width(slice_at_char_boundaries(source, range)),
+            1,
+        ) {
```

Comment:
Perhaps this would be nicer:
```rust
        let space_source = slice_at_char_boundaries(source, 0..range.start);
        let space_len = self.config.width(&space_source);
        write!(self, " {space: >width$}", space = "", width = space_len)?;
        self.set_color(self.styles().label(severity, label_style))?;

        let source = slice_at_char_boundaries(source, range);
        // We use `usize::max` here to ensure that we print at least one
        // label character - even when we have a zero-length span.
        for _ in 0..usize::max(self.config.width(&source), 1) {
```
Sometimes you need to work with rustfmt to get it to format nicely! :) :)